src/LeafInputSignature.cpp

```diff
@@ -0,0 +1,192 @@
+    return !IsFileGenerated((DagRuntimeData*)_userData, Djb2HashPath(includedFile), includedFile);
+}
+
+HashDigest ComputeLeafInputSignature(const Frozen::Dag* dag, const Frozen::DagDerived* dagDerived, const DagRuntimeData *dagRuntime, const Frozen::DagNode* dagNode, MemAllocHeap* heap, MemAllocLinear* scratch, int profilerThreadId, StatCache* stat_cache, DigestCache* digest_cache, ScanCache* scan_cache, FILE* ingredient_stream)
+{
+    ProfilerScope profiler_scope("ComputeLeafInputSignature", profilerThreadId, dagNode->m_Annotation);
+
+    HashState hashState;
+    HashInit(&hashState);
+
+    HashUpdate(&hashState, &dagDerived->m_LeafInputHash_Offline, sizeof(HashDigest));
+
+    const FrozenArray<FrozenFileAndHash>& leafInputs = dagDerived->m_NodeLeafInputs[dagNode->m_DagNodeIndex];
+
+    HashSet<kFlagPathStrings> explicitLeafInputs, implicitLeafInputs;
+    HashSetInit(&explicitLeafInputs, heap);
+    HashSetInit(&implicitLeafInputs, heap);
+    for (auto& leafInput: leafInputs)
+        HashSetInsert(&explicitLeafInputs, leafInput.m_FilenameHash, leafInput.m_Filename.Get());
+
+    ScanInput scanInput;
+    scanInput.m_ScanCache = scan_cache;
+    scanInput.m_ScratchAlloc = scratch;
+    scanInput.m_ScratchHeap = heap;
+
+    IncludeFilterCallback ignoreCallback;
+    ignoreCallback.userData = (void*)dagRuntime;
+    ignoreCallback.callback = FilterOutGeneratedIncludedFiles;
+
+    for (const Frozen::ScannerIndexWithListOfFiles& scannerIndexWithListOfFiles : dagDerived->m_Nodes_to_ScannersWithListsOfFiles[dagNode->m_DagNodeIndex])
+    {
+        scanInput.m_ScannerConfig = dag->m_Scanners[scannerIndexWithListOfFiles.m_ScannerIndex];
+        for (const FrozenFileAndHash& file: scannerIndexWithListOfFiles.m_FilesToScan)
+        {
+            scanInput.m_FileName = file.m_Filename;
+            ScanOutput scanOutput;
+            if (ScanImplicitDeps(stat_cache, &scanInput, &scanOutput, &ignoreCallback))
+            {
+                for(int i=0; i < scanOutput.m_IncludedFileCount; i++)
+                {
+                    const FileAndHash& includedFile = scanOutput.m_IncludedFiles[i];
+                    if (HashSetLookup(&explicitLeafInputs, includedFile.m_FilenameHash, includedFile.m_Filename))
+                        continue;
+                    if (HashSetLookup(&implicitLeafInputs, includedFile.m_FilenameHash, includedFile.m_Filename))
+                        continue;
+                    HashSetInsert(&implicitLeafInputs, includedFile.m_FilenameHash, includedFile.m_Filename);
+                }
+            }
+        }
+    }
+
+    auto addFileContentsToHash = [&](const char* filename, uint32_t filename_hash, const char* label)
+    {
+        HashDigest digest = ComputeFileSignatureSha1(stat_cache, digest_cache, filename, filename_hash);
+
+        if (ingredient_stream)
+        {
+            char digestString[kDigestStringSize];
+            DigestToString(digestString, digest);
+            char buffer[1000];
```

Comment:
Let's not use arbitrary string sizes here. If we want a fixed size, let's use `kMaxPathLength`.

Reply:
fixed